### PR TITLE
This adds typesafe FormControls and FCGroups.

### PR DESCRIPTION
**Pull request by malaupa**
BaseFormControlFactory:

* is not copied in every api module anymore, dependency on
openapi-typescript-angular-generator is used instead
* openapi-typescript-angular-generaor should be install as dependency on
projects to work correctly
* createFormControl now accpets custom validators and updatedOn config

TypedFormControl:

* adds typesafe methods
* generates list with added validators to read by TypedFormControlGroup

TypedFormControlGroup:

* hasControlErrors method detects errors for given form control name
* isValidatorRegistered method detects if a validator function has been
set for form control name
* nextControlErrorKey method returns a generated key for given form
control name
* dispatchErrors method sets errors to child controls based on key in
errors map

Build system with rollup is added. Tests added.

### CODE REVIEW

File: src/formgroup.ts [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+  getNestedGroup<K>(path: Extract<keyof T, string>): TypedFormGroup<K> | null {
+    const control = this.get(path);
+    if (control instanceof TypedFormGroup) {
+      return control;
+    }
+    return null;
+  }
+
+  /**
+   * Detects if a error is present for given control name.
```
Comment (amjmhs) [inline src/formgroup.ts]: Typo: ... a**n** error ...

File: src/formgroup.ts [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+   *
+   * @param name control name of the form group
+   * @param validatorName validator name
+   */
+  isValidatorRegistered(name: Extract<keyof T, string>, validatorName: string): boolean {
+    return this.registeredValidatorsMap[name].some(errorKey => errorKey === validatorName);
+  }
+
+  /**
+   * Returns a error key for the next error (<controlName>.<errorKey>).
```
Comment (amjmhs) [inline src/formgroup.ts]: Typo: ... a**n** error ...

File: src/formgroup.ts [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+      }
+      if (error) {
+        return `${prefix ? `${prefix}.` : ''}${name}.${error}`;
+      }
+    }
+    return '';
+  }
+
+  /**
+   * Dispatches errors this control and to child controls using given error map.
```
Comment (amjmhs) [inline src/formgroup.ts]: Typo? ... "Dispatches errors **to** this control..."

File: src/formgroup.ts [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+   *
+   * @param errors error map
+   * @param contextPath optional context path to errors set to
+   */
+  dispatchErrors(errors: { [key: string]: ValidationErrors }, contextPath?: string) {
+    const paths = Object.keys(errors);
+    paths.forEach(path => {
+      const control = this.get(<Extract<keyof T, string>>(contextPath ? `${contextPath}.${path}` : path));
+      if (control) {
+        control.markAsTouched();
```
Comment (amjmhs) [inline src/formgroup.ts]: Why do you "touch" there? Maybe you can add a comment with the reason ?

File: src/mustache/modelGenericValidators.mustache [hunk 5]
```diff
@@ -4,27 +4,27 @@
 /**
  * A map of `ValidatorFn` for each property of {{classname}}.
```
Comment (amjmhs) [inline src/mustache/modelGenericValidators.mustache]: Documentation needs to be adjusted.

File: src/baseformcontrol.factory.ts [hunk 6]
```diff
@@ -37,14 +38,20 @@ export class BaseFormControlFactory<T> {
...
     const model = this.map.get(property);
-    return new FormControl(model.value, model.validators, asyncValidator);
+    if (model) {
+      return new TypedFormControl(model.value, {
+        validators: [...model.validators, ...(controlOpts ? controlOpts.validators : [])],
+        asyncValidators: controlOpts ? controlOpts.asyncValidators : undefined,
+        updateOn: controlOpts ? controlOpts.updateOn : undefined,
+      });
+    }
+    return new TypedFormControl();
```
[the remark this answers is not in the document]
Comment (malaupa) [reply src/baseformcontrol.factory.ts]: If the method is used in a non typesafe manner.

File: src/formcontrol.ts [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+  private generateValidatorNames(validatorOpts: ValidatorOptions | undefined) {
+    let keys: string[] = [];
+    if (validatorOpts) {
+      let validatorsList = [validatorOpts.validators];
+      if (validatorOpts.asyncValidators) {
+        validatorsList.push(validatorOpts.asyncValidators);
+      }
+      validatorsList.forEach((validators: [string, ValidatorFn | AsyncValidatorFn][]) => {
+        keys = [
+          ...keys,
```
[the remark this answers is not in the document]
Comment (malaupa) [reply src/formcontrol.ts]: It should be an concat not a push of an array into an array.

File: src/formcontrol.ts [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+          ...validators
+            .map(validator => validator && validator[0])
+            // filter duplicates
+            .filter((key, index, array) => array.indexOf(key) === index),
```
[the remark this answers is not in the document]
Comment (malaupa) [reply src/formcontrol.ts]: order is lost in your example

File: src/formcontrol.ts [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+      asyncValidators:
+        opts && opts.asyncValidators ? opts.asyncValidators.map(validator => validator && validator[1]) : null,
+      updateOn: opts && opts.updateOn ? opts.updateOn : 'change',
+    });
+
+    this.registeredValidators = this.generateValidatorNames(opts);
+  }
+
+  /** @inheritdoc */
+  patchValue(
```
[the remark this answers is not in the document]
Comment (malaupa) [reply src/formcontrol.ts]: For type safty...

**Pull request merged**